src/Chapter3.hs

```diff
@@ -480,7 +532,81 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
-{-
+data Castle
+    = Castle String
+    | NoCastle
+
+data Wall
+    = Wall
+    | NoWall
+
+-- tried to give both Wall and Castle a None alternative but the compilare complained
+-- about multiple declarations of None, is there any way to bypass this ?
+
+data Center
+    = Church
+    | Library
+
+data House = MkHouse
+    { people  :: Int }
+
+data City = MkCity
+    { cityCastle  :: Castle
+    , cityWall    :: Wall
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable


 You already have `Castle` that stores walls inside. 

Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```